sha/host/main.c

```diff
@@ -0,0 +1,370 @@
+	op.params[1].value.a = key_size;
+	op.params[2].value.a = obj_type;
+
+	res = TEEC_InvokeCommand(&ctx->sess, TA_SHA_CMD_PREPARE,
+				 &op, &origin);
+	if (res != TEEC_SUCCESS)
+		errx(1, "TEEC_InvokeCommand(PREPARE) failed 0x%x origin 0x%x",
+			res, origin);
+}
+
+void set_key(struct test_ctx *ctx, char *key, size_t key_sz)
+{
+	TEEC_Operation op = {0};
+	uint32_t origin = 0;
+	TEEC_Result res = TEEC_ERROR_GENERIC;
+
+	op.paramTypes = TEEC_PARAM_TYPES(TEEC_MEMREF_TEMP_INPUT,
+					 TEEC_NONE, TEEC_NONE, TEEC_NONE);
+
+	op.params[0].tmpref.buffer = key;
+	op.params[0].tmpref.size = key_sz;
+
+	res = TEEC_InvokeCommand(&ctx->sess, TA_SHA_CMD_SET_KEY,
+				 &op, &origin);
+	if (res != TEEC_SUCCESS)
+		errx(1, "TEEC_InvokeCommand(SET_KEY) failed 0x%x origin 0x%x",
+			res, origin);
+}
+
+void set_iv(struct test_ctx *ctx, char *iv, size_t iv_sz)
+{
+	TEEC_Operation op = {0};
+	uint32_t origin = 0;
+	TEEC_Result res = TEEC_ERROR_GENERIC;
+
+	op.paramTypes = TEEC_PARAM_TYPES(TEEC_MEMREF_TEMP_INPUT,
+					  TEEC_NONE, TEEC_NONE, TEEC_NONE);
+	op.params[0].tmpref.buffer = iv;
+	op.params[0].tmpref.size = iv_sz;
+
+	res = TEEC_InvokeCommand(&ctx->sess, TA_SHA_CMD_SET_IV,
+				 &op, &origin);
+	if (res != TEEC_SUCCESS)
+		errx(1, "TEEC_InvokeCommand(SET_IV) failed 0x%x origin 0x%x",
+			res, origin);
+}
+
+void sha_update_ops(struct test_ctx *ctx, void *message, size_t message_sz,
+		    void *hmac_buff, size_t *hmac_sz)
+{
+	TEEC_Operation op = {0};
+	uint32_t origin = 0;
+	TEEC_Result res = TEEC_ERROR_GENERIC;
+
+	op.paramTypes = TEEC_PARAM_TYPES(TEEC_MEMREF_TEMP_INPUT,
+					 TEEC_MEMREF_TEMP_OUTPUT,
+					 TEEC_NONE, TEEC_NONE);
+
+	op.params[0].tmpref.buffer = message;
+	op.params[0].tmpref.size = message_sz;
+	op.params[1].tmpref.buffer = hmac_buff;
+	op.params[1].tmpref.size = *hmac_sz;
+
+	res = TEEC_InvokeCommand(&ctx->sess, TA_CMD_SHA_INIT, &op,
+				 &origin);
+
+	*hmac_sz = op.params[1].tmpref.size;
+
+	if (res != TEEC_SUCCESS) {
+		errx(1, "TEEC_InvokeCommand(SHA_OPS) failed 0x%x origin 0x%x",
+		     res, origin);
+	}
+}
+
+TEEC_Result compare_hmac_sha(struct test_ctx *ctx, void *message, size_t message_sz,
+		      void *hmac_buff, size_t *hmac_sz)
+{
+	TEEC_Operation op = {0};
+	uint32_t origin = 0;
+	TEEC_Result res = TEEC_ERROR_GENERIC;
+
+	op.paramTypes = TEEC_PARAM_TYPES(TEEC_MEMREF_TEMP_INPUT,
+					 TEEC_MEMREF_TEMP_INPUT,
+					 TEEC_NONE, TEEC_NONE);
+
+	op.params[0].tmpref.buffer = message;
+	op.params[0].tmpref.size = message_sz;
+	op.params[1].tmpref.buffer = hmac_buff;
+	op.params[1].tmpref.size = *hmac_sz;
+
+	res = TEEC_InvokeCommand(&ctx->sess, TA_CMD_SHA_CMPR, &op,
+				 &origin);
+
+	return res;
+}
+
+int main(int argc, char *argv[])
+{
+	struct test_ctx ctx = {0};
+	size_t key_size = 128;
+	void *message = NULL;
+	size_t message_sz = 0;
+	char buff[64] = {0};
+	size_t buff_sz = sizeof(buff);
+	char *algo = NULL;
+	uint32_t obj_type = TA_TYPE_HMAC_SHA256;
+	uint32_t algo_type = SHA_HMAC;
```

Comment:
The algo type is internally used by the CA (host application), it's not used by the TA. There is no need the 2 IDs are defined in the TA API header file. I suggest you only define them in the scope of this **main.c** source file.